### PR TITLE
Add Empty state components and use them

### PR DESCRIPTION
**Pull request by ZitaNemeckova**
How it looks so far:
Screenshot 2021-01-26 at 13 53 44
Screenshot 2021-01-26 at 13 52 30
Screenshot 2021-01-26 at 14 49 05
Screenshot 2021-01-26 at 14 51 27

Screenshot 2021-01-26 at 16 09 37

Screenshot 2021-01-26 at 16 20 05

Screenshot 2021-01-25 at 15 56 28

TODO: 
- [x] add EmptyState component for "filter", "unauthorised" and "no data yet"
- [x] make them fully functional
  - [x] Clear all filters - removed based on UX feedback
  - [x] Add button 
- [x] replace code with components
  - [x] filter
  - [x] unauthorised
  - [x] no data yet
- [x] add check for filter/no data
- [x] src/containers/collection-detail/collection-docs.tsx -> Use EmptyState with red ! icon - followup PR
- [x] EmptyStateNoData -> if possible to add data use "add" icon + primary button else use "search" icon
- [x] "No stuff yet" -> no headers, no footer, no pagination, just title
- [x] fix redirect as it's not always `not-found`
- [x] Exe. Env. needs to use the new nes too.

List of all pages:

Empty state - No data yet:
- Collections
- Namespaces
- My namespaces
- Users (it's there but it shouldn't happen)
- Groups
- Groups - select a group - click on Users tab
- Repo management - Remote tab (it's there but it shouldn't happen)
- Approval 

Empty state - No data for filter:
- Collections
- Namespaces
- My namespaces
- Users
- Groups
- Groups - select a group - click on User tab
- Approval

Empty state - Unauthorised:
- Groups (view and add/edit)
- Users (view and add/edit)
- Approval

Custom "errror" state:
- Collections - select a collection - click documentation tab (it's there but it shouldn't happen)

### CODE REVIEW

**Comment by ZitaNemeckova** [comment]
@sbuenafe-rh @fionayhlin does it look to be heading the right way?

**Comment by sbuenafe-rh** [comment]
@ZitaNemeckova for the **No users added to this group**, you can shorten that title to **No users**

**Comment by fionayhlin** [comment]
@ZitaNemeckova @sbuenafe-rh Should that be **No users yet** to match the other titles?

**Comment by sbuenafe-rh** [comment]
@fionayhlin @ZitaNemeckova yes! good catch.

**Comment by ZitaNemeckova** [comment]
Container Registry:
Screenshot 2021-02-03 at 12 39 09
Screenshot 2021-02-03 at 12 39 52

Groups:
Screenshot 2021-02-03 at 12 48 18

Collection documentation:
Screenshot 2021-02-03 at 12 50 46

Group detail - users:
Screenshot 2021-02-03 at 12 54 21

@sbuenafe-rh @fionayhlin here's new Container Registry empty pages and fixed ones. WDYT?

**Comment by ZitaNemeckova** [comment]
@himdel please have a look code-wise, thanks :)

**Comment by himdel** [comment]
Sure, thanks for the help getting it working :), some testing...

Repo Management > Remote tab - always shows empty state, on master there are 2 items
My namespaces - Create button doesn't work (the empty state one, the filter one does, but shouldn't be there)
Namespace & My namespaces - filter still there on empty state

(I have yet to go through the code, next comment :))

(I also noticed the UI doesn't stop me from removing the current user, which is probably a bug too, though unrelated :))

And do we have a list of all the screens with an EmptyState?

File: src/components/index.ts [hunk 1]
```diff
@@ -48,3 +48,7 @@ export { RemoteRepositoryTable } from './repositories/remote-repository-table';
 export { LocalRepositoryTable } from './repositories/local-repository-table';
 export { StatusIndicator } from './status/status-indicator';
 export { HelperText } from './helper-text/helper-text';
+export { EmptyStateFilter } from './empty-state/empty-state-filter';
+export { EmptyStateUnauthorised } from './empty-state/empty-state-unauthorised';
```
Comment (himdel) [inline src/components/index.ts:52]: this should probably be spelled Unauthori**z**ed (`s` -> `z`)?

File: src/components/repositories/remote-repository-table.tsx [hunk 2]
```diff
@@ -69,28 +55,12 @@ export class RemoteRepositoryTable extends React.Component<IProps> {
 
   render() {
     const { remotes } = this.props;
-    if (remotes.length == 0) {
+    if (remotes.length !== 0) {
```
Comment (himdel) [inline src/components/repositories/remote-repository-table.tsx:58]: this is causing the Remote Repositories issue

**Comment by chouseknecht** [comment]
@himdel Thanks for reviewing our code! Appreciate the comments. I think @alikins is working on the issue regarding users deleting their own account. [View the issue here](https://issues.redhat.com/browse/AAH-265).

**Comment by ZitaNemeckova** [comment]
Screenshot 2021-02-08 at 12 52 07
Screenshot 2021-02-08 at 13 18 14
Screenshot 2021-02-08 at 13 18 24
Screenshot 2021-02-08 at 13 28 08
Screenshot 2021-02-08 at 13 28 18
Screenshot 2021-02-08 at 13 56 32

@sbuenafe-rh @fionayhlin some more Empty State pages. WDYT?

**Comment by fionayhlin** [comment]
@ZitaNemeckova @sbuenafe-rh I believe the no results found from search keyword ones should have a "Clear all filters" link button in the empty state under the text, but otherwise looks good!

**Comment by ZitaNemeckova** [comment]
@fionayhlin We agreed with @sbuenafe-rh that filter handling should stay in filter component only for this case
Screenshot 2021-02-03 at 12 39 09 I assume it included all searches. WDYT?

**Comment by ZitaNemeckova** [comment]
@himdel It's ready for another round of review. Thanks :)

File: src/components/repositories/remote-repository-table.tsx [hunk 3]
```diff
@@ -69,30 +55,6 @@ export class RemoteRepositoryTable extends React.Component<IProps> {
 
   render() {
     const { remotes } = this.props;
-    if (remotes.length == 0) {
```
Comment (himdel) [inline src/components/repositories/remote-repository-table.tsx:72]: :+1: handled by `repository-list` now, the only user of `RemoteRepositoryTable`

**Comment by himdel** [comment]
I think `EmptyStateUnauthorised` is currently missing a screenshot, and unless it's supposed to be different..

```diff
diff --git a/src/components/empty-state/empty-state-unauthorised.tsx b/src/components/empty-state/empty-state-unauthorised.tsx
index a205ca6..bef7e2e 100644
--- a/src/components/empty-state/empty-state-unauthorised.tsx
+++ b/src/components/empty-state/empty-state-unauthorised.tsx
@@ -15,9 +15,9 @@ interface IProps {}
 export class EmptyStateUnauthorised extends React.Component<IProps> {
   render() {
     return (
-      <EmptyState variant={EmptyStateVariant.xs}>
+      <EmptyState variant={EmptyStateVariant.small}>
         <EmptyStateIcon icon={LockIcon} />
-        <Title headingLevel='h4'>
+        <Title headingLevel='h4' size='lg'>
           You do not have have access to Automation Hub
         </Title>
         <EmptyStateBody>
```

(And I still think Unauthori**z**ed is the proper spelling, unless Ansible is using non-US spellings?)

---

Would it make sense to DRY it up by implementing the other empty states on top of `EmptyStateCustom`?

If we move the `{this.props.button && (...` logic there, all the other empty states could just be built on top of it, like..

```js
export const EmptyStateFilter = ({}) => (<EmptyStateCustom
  title="No resuls found"
  description="No results match the filter criteria. Remove all filters or clear all filters to show results."
  icon={SearchIcon}
/>);

export const EmptyStateNoData = ({ title, description, button }) => (<EmptyStateCustom
  title={title}
  description={description}
  icon={button ? PlusCircleIcon : SearchIcon}
  button={button}
/>);

export const EmptyStateUnauthorized = ({}) => (<EmptyStateCustom
  title="You do not have have access to Automation Hub"
  description="Contact you organization administrator for more information."
  icon={LockIcon}
/>);
```

(+- typescript and the component syntax anyway)

---

I'm seeing some icon changes, `InfoIcon` -> `SearchIcon`, `WarningTriangleIcon` -> `SearchIcon`, but those are all intentional, right? :)

And I'll continue tomorrow :)

File: src/utilities/filter-is-set.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+// Checks that at least one filter is set
+export function filterIsSet(params, filters) {
```
Comment (himdel) [inline src/utilities/filter-is-set.ts:2]: This could be `_.some(filters, (filter) => filter in params));`, if lodash is preferred.

**Comment by ZitaNemeckova** [comment]
@himdel Icon changes are intentional based on UX requirements. `s/z` is fixed. I'll do the DRY you suggested as it's really cool! 

Unathorised:
Screenshot 2021-02-17 at 11 46 01

File: src/utilities/filter-is-set.ts (hunk 4 above)
Comment (ZitaNemeckova) [reply src/utilities/filter-is-set.ts:3]: Fixed :)

**Comment by ZitaNemeckova** [comment]
@himdel last review addressed. Ready for another round. Thanks :)

File: src/containers/namespace-list/namespace-list.tsx [hunk 5]
```diff
@@ -140,48 +135,65 @@ export class NamespaceList extends React.Component<IProps, IState> {
...
-              sortOptions={[{ title: 'Name', id: 'name', type: 'alpha' }]}
-              searchPlaceholder={'Search ' + title.toLowerCase()}
-              updateParams={p =>
-                this.updateParams(p, () => this.loadNamespaces())
-              }
-              extraInputs={extra}
-            />
-            <div>
-              <Pagination
+          {filterIsSet(this.state.params, ['keywords']) ||
```
Comment (himdel) [inline src/containers/namespace-list/namespace-list.tsx:138]: consider using a `noData` var as well?

File: src/containers/repositories/repository-list.tsx [hunk 6]
```diff
@@ -188,36 +185,50 @@ class RepositoryList extends React.Component<RouteComponentProps, IState> {
...
+      ) : (
+        <Main className='repository-list'>
+          <Section className='body'>
+            {loading ? (
+              <LoadingPageSpinner />
+            ) : (
+              <RemoteRepositoryTable
+                remotes={content}
+                updateParams={this.updateParams}
+                editRemote={(remote: RemoteType) => {
```
Comment (himdel) [inline src/containers/repositories/repository-list.tsx:217]: `this.selectRemoteToEdit` becomes unused (and does the same thing except on a *copy* of `remote`?)

File: src/containers/user-management/user-create.tsx [hunk 7]
```diff
@@ -31,17 +36,27 @@ class UserCreate extends React.Component<RouteComponentProps, IState> {
 
   render() {
     const { user, errorMessages } = this.state;
-    if (!this.context.user || !this.context.user.model_permissions.add_user) {
-      return <Redirect to={Paths.notFound}></Redirect>;
-    }
-    return (
+    const notAutthorised =
```
Comment (himdel) [inline src/containers/user-management/user-create.tsx:39]: typo, double T

**Comment by himdel** [comment]
Thanks, looks great :)

---

Is `redirect` ever set from outside the component, or can I assume `setState({ redirect: ...` only happens in the component itself?

If it's local only, would it make sense to split redirect from 404 and use a separate state param? It looks like some components (`edit-namespace`) still need to Redirect after update, so you're catching `Paths.notFound` as a special case and dealing with it differently, while others (`certification-dashboard`) only used `Paths.notFound` so the `Redirect` is removed and `redirect` is used to just mean "not there".

I'm not seeing any logic errors in the conversion :+1: , but if it's safe to split, I'd keep `redirect` just for `Redirect` and use something like `setState({ notFound: true })` instead for the unauthorized empty state.

EDIT: or `unauthorized` like in `user-edit`

---

I do notice that `filterIsSet` needs to know the list of params that *could* be filters while `AppliedFilters` seems to need a list of params to *ignore*.. sounds like the 2 lists could get out of sync fairly quickly, new params would get ignored by `filterIsSet` but not by `AppliedFilters`, we may want to fix that by having a per screen list of filers, used by both.

And perhaps the meta params could be app global, enforcing `page_size`, `id` etc. can never be treated as a filter, regardles of whether the screen is using it or not (and we could drop the `ignoredParams` list). But that's probably a separate refactoring as that's not really related to this change?

---

I can't seem to be able to rebuild now (`pkg_resources.VersionConflict: (pulp-container 2.1.0 (/venv/lib/python3.6/site-packages), Requirement.parse('pulp-container==2.3.1'))`, possibly related to https://github.com/ansible/galaxy_ng/pull/650 ?), so I have not tested the new changes yet.

File: src/containers/user-management/user-create.tsx (hunk 7 above)
Comment (ZitaNemeckova) [reply src/containers/user-management/user-create.tsx:39]: fixed :)

**Comment by ZitaNemeckova** [comment]
@himdel

`redirect` shouldn't be set from outside. I'll fix it as suggested.

Params refactoring sounds like good idea but it's out of scope for this PR.

How to rebuild:  `galaxy_ng` must be on latest `master`, `LOCK_REQUIREMENTS=1` added  in `.compose.env` then build and run migration (to be safe).

File: src/containers/namespace-list/namespace-list.tsx (hunk 5 above)
Comment (ZitaNemeckova) [reply src/containers/namespace-list/namespace-list.tsx]: Added `noData` var. Leaving 209 as is.

File: src/containers/repositories/repository-list.tsx (hunk 6 above)
Comment (ZitaNemeckova) [reply src/containers/repositories/repository-list.tsx]: I cannot remember why I did that. Fixing back at `this.selectRemoteToEdit`

File: src/containers/namespace-list/namespace-list.tsx [hunk 8]
```diff
@@ -109,6 +104,8 @@ export class NamespaceList extends React.Component<IProps, IState> {
     const { namespaces, params, itemCount } = this.state;
     const { title, filterOwner } = this.props;
     const { user } = this.context;
+    const noData =
+      !filterIsSet(this.state.params, ['keywords']) && namespaces.length === 0;
```
Comment (himdel) [inline src/containers/namespace-list/namespace-list.tsx:108]: `namespaces` can be undefined, we can't check length if it is

**Comment by himdel** [comment]
Ah, managed to rebuild, `docker system prune -a --volumes` helped, seems like the build doesn't notice changes to requirements :).

---

Trying to load `/ui/namespaces` or `/ui/my-namespaces` (with a fresh db):

```
 Cannot read property 'length' of undefined
    at NamespaceList../src/containers/namespace-list/namespace-list.tsx.NamespaceList.render (App.js:12364)
```

(https://github.com/ansible/ansible-hub-ui/pull/285#discussion_r578834943)

---

`src/containers/certification-dashboard/certification-dashboard.tsx` is still using the `redirect: Paths.notFound` pattern to display `EmptyStateUnauthorized`

`src/containers/namespace-detail/namespace-detail.tsx` is still redirecting (actual redirect) to `Paths.notFound`, should it be doing that?

File: src/containers/namespace-list/namespace-list.tsx (hunk 8 above)
Comment (ZitaNemeckova) [reply src/containers/namespace-list/namespace-list.tsx]: Fixed :+1:

**Comment by ZitaNemeckova** [comment]
@himdel
> `src/containers/namespace-detail/namespace-detail.tsx` is still redirecting (actual redirect) to Paths.notFound, should it be doing that?

Yes, it gets all `http://localhost:8002/ui/<ramdom stuff>` calls. This my be worth refactoring later.

**Review by himdel — APPROVED** [review]
LGTM, not seeing any issues in code or when testing :+1:

**Pull request merged**